### PR TITLE
feat: apply gateway lookback limit to eth API lookback

### LINKED ISSUE #10412: Gateway: EthAPI needs to restrict lookbacks
The gateway currently supports some EthAPI methods that could have unrestricted lookbacks. Upon examination, I think the only such methods are:

- EthGetTransactionReceipt
- EthGetTransactionByHash

If called with tx hashes from far in the past, these methods will walk up the chain potentially as far as genesis looking for the relevant message. Due to an inefficiency in `EthGetTransactionReceipt`, it will actually perform this walkup twice, but that is unrelated and easily resolved.

There's an open question about what we want to do about this. We could:

- Leave it as-is: Any good gateway will have a short timeout anyway, the requests will just fail ([default is 5 seconds](https://github.com/filecoin-project/lotus/blob/8975f0b753085854a84dec0049554cad7fdd0c4e/gateway/node.go#L35)).
- Drop these methods from the gateway: I'm guessing we don't want to do this, but it's an option.
- Refactor the system to use the "regular" gateway lookback limits. This is possible, but requires some code moving around. I would propose doing it as follows (not the fastest, but IMO the cleanest way to do this).
  - Add an `EthLookupEthAddress` method to the API. I think we should have this regardless.
  - Refactor `newEthTxFromMessageLookup`, etc. methods to be moved to the `types` package (perhaps defined on the types themeselves). Doing so will require replacing the StateAPI param these methods take with a `lookupEthAddr` func that callers can supply.
  - Extract out the "[get message lookup](https://github.com/filecoin-project/lotus/blob/12496e56317377ece4af744f065d6975ab82674c/node/impl/full/eth.go#L277-L294)" pattern that given a tx hash retrieves the message lookup for it. Implement this in the gateway impl with lookbacks on StateSearchMsg.
  - Modify the gateway impls to use the logic extracted in the previous step. This effectively reimplements the methods on the gateway, but I think can be 2-3 line methods that mostly rely on logic in the types packages.

### PR DESCRIPTION
**Pull request by Stebalien**
## Related Issues

fixes #10412

## Proposed Changes

This change applies the gateway lookback limits to the eth txn and receipt lookup APIs. It:

1. Introduces new "limited" API endpoints for EthGetTransactionByHash and EthGetTransactionReceipt that accept lookback-limits.
2. Implements the gateway version of these API endpoints by calling the limited variants with the default message search lookback limit.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] New features have usage guidelines and / or documentation updates in
  - [x] [Lotus Documentation](https://lotus.filecoin.io)
  - [x] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [x] CI is green

### CODE REVIEW

**Comment by magik6k** [comment]
Gateway tests are angry
```
--- FAIL: TestGatewayWalletMsig (11.08s)
panic: reflect.Set: value of type *api.Gateway is not assignable to type full.EthModuleAPI [recovered]
	panic: reflect.Set: value of type *api.Gateway is not assignable to type full.EthModuleAPI

goroutine 71 [running]:
testing.tRunner.func1.2({0x447bb40, 0xc00c58fdf0})
	/usr/local/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1392 +0x39f
panic({0x447bb40, 0xc00c58fdf0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
reflect.Value.assignTo({0x43d1520?, 0xc00c58fde0?, 0xc00111d998?}, {0x4a82910, 0xb}, 0x4915b20, 0xc00c58fdd0)
	/usr/local/go/src/reflect/value.go:3064 +0x2ac
reflect.Value.Set({0x4915b20?, 0xc00c58fdd0?, 0x4915b20?}, {0x43d1520?, 0xc00c58fde0?, 0x1?})
	/usr/local/go/src/reflect/value.go:2090 +0xeb
github.com/filecoin-project/lotus/node.as.func2({0xc00679ef48?, 0x2?, 0x2?})
	/home/circleci/lotus/node/options.go:156 +0x233
reflect.Value.call({0xc00a0bde00?, 0xc00d04f200?, 0x7574e5?}, {0x4a4cb74, 0x4}, {0xc00679ef30, 0x1, 0x30?})
	/usr/local/go/src/reflect/value.go:556 +0x845
reflect.Value.Call({0xc00a0bde00?, 0xc00d04f200?, 0x757847?}, {0xc00679ef30, 0x1, 0x1})
	/usr/local/go/src/reflect/value.go:339 +0xbf
go.uber.org/dig.defaultInvoker({0xc00a0bde00?, 0xc00d04f200?, 0xc00d408f60?}, {0xc00679ef30?, 0x1?, 0x59970b0?})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/dig.go:355 +0x28
go.uber.org/dig.(*node).Call(0xc00d403180, {0x59970b0?, 0xc00d01b7c0})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/dig.go:806 +0x259
go.uber.org/dig.paramSingle.Build({{0x0, 0x0}, 0x0, {0x59a7c48, 0x4915b20}}, {0x59970b0, 0xc00d01b7c0})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/param.go:245 +0x242
go.uber.org/dig.paramObjectField.Build(...)
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/param.go:401
go.uber.org/dig.paramObject.Build({{0x59a7c48, 0x49f16a0}, {0xc000192fa0, 0x3, 0x4}}, {0x59970b0, 0xc00d01b7c0})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/param.go:328 +0x148
go.uber.org/dig.paramObjectField.Build(...)
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/param.go:401
go.uber.org/dig.paramObject.Build({{0x59a7c48, 0x4a20600}, {0xc00d3fd180, 0x10, 0x10}}, {0x59970b0, 0xc00d01b7c0})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/param.go:328 +0x148
go.uber.org/dig.paramList.BuildList({{0x59a7c48, 0xc000156300}, {0xc00e71dea0, 0x1, 0x1}}, {0x59970b0, 0xc00d01b7c0})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/param.go:201 +0xb9
go.uber.org/dig.(*Container).Invoke(0xc00d01b7c0, {0xc000156300?, 0xc00d02e300}, {0x1b5665a?, 0x1?, 0x1?})
	/home/circleci/go/pkg/mod/go.uber.org/dig@v1.12.0/dig.go:503 +0x2b9
go.uber.org/fx.(*App).executeInvoke(0xc00d07ec30, {{0xc000156300, 0xc00d02e300}, {0xc00d022c80, 0x7, 0x8}})
	/home/circleci/go/pkg/mod/go.uber.org/fx@v1.15.0/app.go:964 +0x39f
go.uber.org/fx.(*App).executeInvokes(...)
	/home/circleci/go/pkg/mod/go.uber.org/fx@v1.15.0/app.go:929
go.uber.org/fx.New({0xc00111f2f8, 0x3, 0x1e?})
	/home/circleci/go/pkg/mod/go.uber.org/fx@v1.15.0/app.go:596 +0xa4b
github.com/filecoin-project/lotus/node.New({0x5983fb0, 0xc00005a068}, {0xc0002741c0, 0xb, 0xe})
	/home/circleci/lotus/node/builder.go:367 +0x477
github.com/filecoin-project/lotus/itests/kit.(*Ensemble).Start(0xc0006d6240)
	/home/circleci/lotus/itests/kit/ensemble.go:456 +0xd3d
command-line-arguments.startNodes({0x5983fb0, 0xc00005a068}, 0xc000703040, 0x8725a0?, 0x0?, 0xc0002765f0?)
	/home/circleci/lotus/itests/gateway_test.go:314 +0x39d
command-line-arguments.TestGatewayWalletMsig(0xd40010?)
	/home/circleci/lotus/itests/gateway_test.go:52 +0x70
testing.tRunner(0xc000703040, 0x55ed0c0)
	/usr/local/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x35f
```

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -88,8 +90,6 @@ type EthEventAPI interface {
 var (
 	_ EthModuleAPI = *new(api.FullNode)
 	_ EthEventAPI  = *new(api.FullNode)
-
-	_ EthModuleAPI = *new(api.Gateway)
```
Comment (arajasek) [inline node/impl/full/eth.go:92]: We should maintain this property, we can do so by implementing the new methods on the `EthAPI` directly as Magik said.

**Review by arajasek — CHANGES_REQUESTED** [review]
The `limit`s aren't actually enforced?

File: node/impl/full/eth.go (hunk 1 above)
Comment (Stebalien) [reply node/impl/full/eth.go:92]: Ah, ok, fair enough.

**Review by magik6k — APPROVED** [review]

**Pull request merged**